authoring-recipes/types-of-recipes.md

```diff
@@ -0,0 +1,88 @@
+# Types of recipes
+
+Before you jump into coding recipes, it's important to figure out what type of recipe you're going to make.
+
+OpenRewrite offers support for three different types of recipes: declarative, refaster, and imperative. Each of these has its own advantages and disadvantages depending on the type of recipe you want to build. 
+
+In this doc, we'll briefly describe each of these so that you can determine what type of recipe works best for your needs. 
+
+## Declarative recipes
+
+[Declarative recipes](/reference/yaml-format-reference.md) are the simplest and most common recipes. They are entirely written in YAML, and they generally tie together existing recipes while adding some light configuration to them.
+
+The [common static analysis recipe](https://github.com/openrewrite/rewrite-static-analysis/blob/main/src/main/resources/META-INF/rewrite/common-static-analysis.yml) is a great example of this. It takes a bunch of simple recipes relating to static analysis and combines them into one larger recipe that can be run more easily. It does not add any logic, and it doesn't change the existing recipes.
+
+```yaml
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.staticanalysis.CommonStaticAnalysis
+displayName: Common static analysis issues
+description: Resolve common static analysis issues discovered through 3rd party tools.
+recipeList:
+  - org.openrewrite.staticanalysis.AtomicPrimitiveEqualsUsesGet
+  - org.openrewrite.staticanalysis.BigDecimalRoundingConstantsToEnums
+  - org.openrewrite.staticanalysis.BooleanChecksNotInverted
+  - org.openrewrite.staticanalysis.CaseInsensitiveComparisonsDoNotChangeCase
+  - org.openrewrite.staticanalysis.CatchClauseOnlyRethrows
+  - org.openrewrite.staticanalysis.ChainStringBuilderAppendCalls
+  # ...
+---
+```
+
+{% hint style="success" %}
+As a best practice, if your recipe can be declarative (meaning it can be built out of other recipes), then you should make it declarative.
+{% endhint %}
+
+## Refaster templates
+
+Refaster templates are the "middle ground" of recipes. They offer more functionality than [declarative recipes](#declarative-recipes), but not as much as [imperative recipes](#imperative-recipes). On the other hand, compared to an imperative recipe, they're much quicker to create and require much less knowledge to get started.
+
+Refaster templates are ideal for straightforward replacements such as converting `StringUtils.equals(..)` to `Objects.equals(..)`. These are more than just a string replacement, though; they offer compiler and type support.
```

Comment:
I think we should try to make it clear that Refaster templates can only be used to replace one expression with another or one statement with another.

Also, we should link to the [Errorprone Refaster documentation](https://errorprone.info/docs/refaster), which isn't very exhaustive, but it still gives a nice overview.

Currently, we only support the `@BeforeTemplate` and `@AfterTemplate` annotations and our own form of the `@Matches` and `@NotMatches` annotations. The latter two aren't even documented in Refaster, except for in the [Javadoc](https://errorprone.info/api/latest/com/google/errorprone/refaster/annotation/package-summary.html). But since we plan to expand on this fairly soon, I don't think we really need to document it.